File: tree/binary-heap.js
class BinaryHeap {
  constructor(type) {
    this.type = type;
    this.values = [];
  }

  /*
    Formula to find the children of a node (n = index of node):
      left child  = (2 * n) + 1
      right child = (2 * n) + 2
  */
  getChildren(index) {
    return {
      leftIndex: index * 2 + 1,
      rightIndex: index * 2 + 2,
    };
  }

  /*
    Formula to find the parent of a node (n = index of node):
      (n - 1) / 2 floored
  */
  getParent(index) {
    return Math.floor((index - 1) / 2);
  }

  insert(val) {
    // Step 1: insert the value to the end of the tree/list
    this.values.push(val);

    // Step 2: compare the new value to it's parent, if it's larger swap the two
    // keep doing this until the new value "bubbles up" to it's rightful place
    let newNodeIndex = this.values.length - 1;
    let parentIndex = this.getParent(newNodeIndex);

    while (
      this.type === "min"
        ? val < this.values[parentIndex]
        : val > this.values[parentIndex]
    ) {
      const parentNode = this.values[parentIndex];
      const newNode = this.values[newNodeIndex];

      this.values[newNodeIndex] = parentNode;
      this.values[parentIndex] = newNode;

      newNodeIndex = parentIndex;
      parentIndex = this.getParent(parentIndex);
    }
  }

  // An `extractMin()` function will be identical except we'd want to get the smaller of the 2 children `Math.min()`
  // and the condition in the while loop will be `smallerChild < curNode`
  extractMax() {
    // Step 1: remove and return the first value (in a max heap the first node is the largest)
    const maxValue = this.values.shift();

    // Step 2: move the last value to the front of the heap temporarily
    this.values.unshift(this.values.pop());

    // Step 3: bubble down the newly inserted last node to it's rightful place until the root node once again holds the largest value
    let curNodeIndex = 0;
    let curNode = this.values[curNodeIndex];

    let children = this.getChildren(curNodeIndex);
    let largerChildIndex = Math.max(children.leftIndex, children.rightIndex); // Since it's a max heap we want the larger of the two
    let largerChild = this.values[largerChildIndex];

    // Swap
    while (largerChild > curNode) {
      this.values[curNodeIndex] = largerChild;
      this.values[largerChildIndex] = curNode;

      curNodeIndex = largerChildIndex;
      curNode = this.values[largerChildIndex];

      children = this.getChildren(curNodeIndex);
      largerChildIndex = Math.max(children.leftIndex, children.rightIndex);
      largerChild = this.values[largerChildIndex];
    }

    return maxValue;
  }

  dfs() {
    const nodes = [];

    const traverse = (index) => {
      nodes.push(this.values[index]);

      const { leftIndex, rightIndex } = this.getChildren(index);

      const leftNode = this.values[leftIndex];
      const rightNode = this.values[rightIndex];

      if (leftNode) traverse(leftIndex);
      if (rightNode) traverse(rightIndex);

      return nodes;
    };

    return traverse(0);
  }
}

console.log("\n<<<-------------- Max Heap ----------->>>\n");

// https://i.imgur.com/SE1ULLs.png

const maxHeap = new BinaryHeap("max");

maxHeap.insert(41);
maxHeap.insert(39);
maxHeap.insert(33);
maxHeap.insert(18);
maxHeap.insert(27);
maxHeap.insert(12);
maxHeap.insert(55); // Bubbles up to the right spot

console.log("Nodes:", maxHeap.values); // [55, 39, 41, 18, 27, 12, 33]

console.log("Dfs:", maxHeap.dfs()); // [55, 39, 18, 27, 41, 12, 33];

console.log("Extract Max:", maxHeap.extractMax()); // 55

console.log("Nodes:", maxHeap.values); // [41, 39, 33, 18, 27, 12]

console.log("\n<<<-------------- Min Heap ----------->>>\n");

// https://i.imgur.com/L1Rn6dO.png

const minHeap = new BinaryHeap("min");

minHeap.insert(2);
minHeap.insert(3);
minHeap.insert(6);
minHeap.insert(5);
minHeap.insert(9);
minHeap.insert(8);
minHeap.insert(1); // Bubbles up to right spot

console.log("Nodes:", minHeap.values); // [1, 3, 2, 5, 9, 8, 6]

console.log("Dfs:", minHeap.dfs()); // [1, 3, 5, 9, 2, 8, 6]

File: tree/binary_heap.py
import math


class BinaryHeap():
    def __init__(self, type):
        self.type = type
        self.values = []

    # Formula to find the parent of a node (n = index of node):
    #   (n - 1) / 2 floored
    def get_parent(self, current_index):
        return math.floor((current_index - 1) / 2)

    # Formula to find the children of a node (n = index of node):
    #   left child  = (n * 2) + 1
    #   right child = (n * 2) + 2
    def get_left_child(self, current_index):
        return current_index * 2 + 1

    def get_right_child(self, current_index):
        return current_index * 2 + 2

    def dfs(self, current_index, nodes=[]):
        nodes.append(self.values[current_index])

        left_child_index = self.get_left_child(current_index)
        right_child_index = self.get_right_child(current_index)

        if left_child_index <= len(self.values) - 1:
            self.dfs(left_child_index)

        if right_child_index <= len(self.values) - 1:
            self.dfs(right_child_index)

        return nodes

    def insert(self, val):
        # Step 1: insert the value to the end of the list
        self.values.append(val)

        # Step 2: keep bubbling the node up until it reaches the right spot
        # starting index is the last elem since we just pushed the new val to the end of the list
        self.bubble_up(len(self.values) - 1)

    def bubble_up(self, current_index):
        if current_index == 0:
            return

        current_node = self.values[current_index]

        parent_index = self.get_parent(current_index)
        parent_node = self.values[parent_index]

        if (self.type == "max" and current_node > parent_node) or (self.type == "min" and current_node < parent_node):
            # Swap
            self.values[current_index] = parent_node
            self.values[parent_index] = current_node

            self.bubble_up(parent_index)

    # This method is only for a max heap
    def extract_max(self):
        # Step 1: remove and return the first value (in a max heap the first node is the largest)
        max_value = self.values.pop(0)

        # If this was the last value in the list there's no need to do Steps 2 or 3
        if len(self.values) == 0:
            return max_value

        # Step 2: move the last value to the front of the heap temporarily
        self.values.insert(0, self.values.pop())

        # Step 3: bubble down the newly inserted last node to it's rightful place until the root node once again holds the largest value
        # starting index is 0 since we just prepended the last item to the front of the list
        self.bubble_down(0)

        return max_value

    def bubble_down(self, current_index):
        current_node = self.values[current_index]

        left_child_index = self.get_left_child(current_index)
        right_child_index = self.get_right_child(current_index)

        left_child = self.values[left_child_index]
        right_child = self.values[right_child_index]

        if left_child > right_child:
            larger_child = left_child
            larger_child_index = left_child_index
        else:
            larger_child = right_child
            larger_child_index = right_child_index

        if larger_child > current_node:
            # Swap
            self.values[current_index] = larger_child
            self.values[larger_child_index] = current_node

            self.bubble_up(larger_child_index)


print("\n<<<-------------- Max Heap ----------->>>\n")

# https://i.imgur.com/SE1ULLs.png

max_heap = BinaryHeap("max")

max_heap.insert(41)
max_heap.insert(39)
max_heap.insert(33)
max_heap.insert(18)
max_heap.insert(27)
max_heap.insert(12)
max_heap.insert(55)  # Bubbles up to the right spot


print("Nodes:", max_heap.values)  # [55, 39, 41, 18, 27, 12, 33]

print("Dfs:", max_heap.dfs(0))  # [55, 39, 18, 27, 41, 12, 33]

print("Extract Max:", max_heap.extract_max())  # 55

print("Nodes:", max_heap.values)  # [41, 39, 33, 18, 27, 12]


print("\n<<<-------------- Min Heap ----------->>>\n")

# https://i.imgur.com/L1Rn6dO.png

min_heap = BinaryHeap("min")

min_heap.insert(2)
min_heap.insert(3)
min_heap.insert(6)
min_heap.insert(5)
min_heap.insert(9)
min_heap.insert(8)
min_heap.insert(1)  # Bubbles up to right spot

print("Nodes:", min_heap.values)  # [1, 3, 2, 5, 9, 8, 6]

print("Dfs:", min_heap.dfs(0))  # [1, 3, 5, 9, 2, 8, 6]
